src/dict_list.jl

```diff
@@ -0,0 +1,210 @@
+calling `dict_list(c)`.
+"""
+function dict_list_count(c)
+    if DependentParameter in eltype.(values(c))
+        return length(dict_list(c))
+    end
+    iterable_fields = filter(k -> typeof(c[k]) <: Vector, keys(c))
+    prod(length(c[i]) for i in iterable_fields)
+end
+
+# Basis taken from https://github.com/FluxML/MacroTools.jl
+walk(x, inner, outer, ex) = outer(x, ex)
+walk(x::Expr, inner, outer, ex) = outer(Expr(x.head, map(y->inner(y,x), x.args)...), ex)
+postwalk(f, x, ex=:()) = walk(x, (x,y) -> postwalk(f, x, y), f, ex)
+
+export @onlyif
+
+struct DependentParameter{T}
+    value::T
+    condition::Function
+end
+
+contains_partially_restricted(d::Dict) = any(contains_partially_restricted,values(d))
+contains_partially_restricted(d::Vector) = any(contains_partially_restricted,d)
+contains_partially_restricted(::DependentParameter) = true 
+contains_partially_restricted(::Any) = false 
+contains_fully_restricted(d::Dict) = any(is_fully_restricted,values(d))
+is_fully_restricted(d::Vector) = eltype(d) <: DependentParameter
+is_fully_restricted(d) = typeof(d) <: DependentParameter
+
+function toDependentParameter(value::T,condition) where T
+    if T <: Vector
+        return DependentParameter.(value,Ref(condition))
+    end
+    return DependentParameter(value,condition)
+end
+
+function lookup_candidate(d, name)
+    if name in keys(d)
+        if d[name] isa DependentParameter
+            return d[name].value
+        end
+        return d[name]
+    end
+    return name
+end
+
+"""
+    @onlyif(ex, value)
+Tag `value` to only appear in a dictionary created with [`dict_list`](@ref) if
+the Julia expression `ex` (see below) is evaluated as true.  If `value` is a subtype of
+`Vector`, `@onlyif` is applied to each entry.
+Since `@onlyif` is applied to a value and not to a dictionary key, it is
+possible to restrict only some of the values of a vector. This means that based
+on `ex` the number of options for a particular key varies.
+
+Within `ex` it is possible to extract values of the dictionary passed to
+[`dict_list`](@ref) by a shorthand notation where only the key must be
+provided.  For example `ex = :(:N == 1)` is tranformed in the call
+`dict_list(d)` to an expression analogous to `:(d[:N] == 1)` by using the
+function [`lookup_candidate`](@ref).  This is supported for `Symbol` and
+`String` keys.
+
+## Examples
+```julia
+julia> d = Dict(:a => [1, 2], :b => 4, :c => @onlyif(:a == 1, [10, 11]));
+
+julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
+3-element Array{Dict{Symbol,Int64},1}:
+ Dict(:a => 1,:b => 4,:c => 10)
+ Dict(:a => 1,:b => 4,:c => 11)
+ Dict(:a => 2,:b => 4)
+
+ julia> d = Dict(:a => [1, 2], :b => 4, :c => [10, @onlyif(:a == 1, 11)]);
+
+julia> dict_list(d) # only in case `:a` is `1` the dictionary will get extra value `11` for key `:c`
+3-element Array{Dict{Symbol,Int64},1}:
+ Dict(:a => 1,:b => 4,:c => 10)
+ Dict(:a => 1,:b => 4,:c => 11)
+ Dict(:a => 2,:b => 4,:c => 10)
+```
```

Comment:
```suggestion
```

See the [Defining parameter sets with restrictions](@ref) section for more examples.

```